api/types.proto

```diff
@@ -713,3 +713,33 @@ message ManagerStatus {
 	// Reachability specifies whether this node is reachable.
 	RaftMemberStatus.Reachability reachability = 4;
 }
+
+// SecretType provides information about what kind of secret this is
+enum SecretType {
+	CONTAINER = 0 [(gogoproto.enumvalue_customname) = "ContainerSecret"];
+	NODE = 1 [(gogoproto.enumvalue_customname) = "NodeSecret"];
+}
+
+// SecretReference is the linkage between a service and a secret that it uses.
+message SecretReference {
+	// Name is the name of the secret that this references.
+	string name = 1;
+
+	// SecretDataID represents the ID of the specific Secret version that we're
+	// referencing. This identifier exists so that SecretReferences don't leak
+	// any information about the secret contents.
+	string secret_data_id = 2 [(gogoproto.customname) = "SecretDataID"];
```

Comment:
> In general, we believe that a hash is irreversible

This assumption is only valid if the data being hashed has very high entropy. It's not generally true for hashes of passwords, which is why compromises of unsalted password hashes very often lead to serious security breaches. I feel like it would be very limiting to put a caveat on Docker secrets that the payloads should never be passwords because the hashes are not treated as confidential. That would be pretty embarrassing.